controllers/machinehealthcheck_controller.go

```diff
@@ -2,61 +2,503 @@ package controllers
+		mhc.Status.RemediationsAllowed = 0
+		utils.SetMachineCondition(mhc, &v1beta1.Condition{
+			Type:     v1beta1.RemediationAllowedCondition,
+			Status:   corev1.ConditionFalse,
+			Severity: v1beta1.ConditionSeverityWarning,
+			Reason:   v1beta1.TooManyUnhealthyReason,
+			Message:  msg,
+		})
+
+		// no need to requeue if we don't have targets, or healthy nodes only
+		if totalTargets == 0 || unhealthyCount == 0 {
+			return reconcile.Result{}, nil
+		}
+
+		return reconcile.Result{Requeue: true}, nil
+	}
+
+	log.Info("Remediations are allowed", "total targets", totalTargets, "max unhealthy", mhc.Spec.MaxUnhealthy, "unhealthy targets", unhealthyCount)
+	metrics.ObserveMachineHealthCheckShortCircuitDisabled(mhc.Name, mhc.Namespace)
+
+	utils.SetMachineCondition(mhc, &v1beta1.Condition{
+		Type:   v1beta1.RemediationAllowedCondition,
+		Status: corev1.ConditionTrue,
+	})
+
+	remediationErrors := r.remediateAll(needRemediation, resourceManager)
+	errList = append(errList, remediationErrors...)
+	if len(errList) > 0 {
+		requeueError := utilerrors.NewAggregate(errList)
+		log.V(3).Info("Reconciling: there were errors, requeuing", "errors", requeueError)
+		return reconcile.Result{}, requeueError
+	}
 
+	if requeueIn > 0 {
+		log.V(3).Info("Reconciling: some targets might go unhealthy. Ensuring a requeue happens", "requeue in", requeueIn.String())
+		return reconcile.Result{RequeueAfter: requeueIn}, nil
+	}
+
+	log.V(3).Info("Reconciling: no more targets meet unhealthy criteria")
 	return result, nil
 }
 
-// SetupWithManager sets up the controller with the Manager.
-func (r *MachineHealthCheckReconciler) SetupWithManager(mgr ctrl.Manager) error {
-	return ctrl.NewControllerManagedBy(mgr).
-		For(&v1beta1.MachineHealthCheck{}).
-		Complete(r)
+func (r *MachineHealthCheckReconciler) checkHealth(targets []resources.Target) (healthy []resources.Target, needRemediation []resources.Target, overallRequeueIn time.Duration, errList []error) {
+
+	for _, t := range targets {
+		log := r.Log.WithValues("target", t.String())
+		log.Info("Reconciling: health checking")
+		needsRemediation, requeueIn, err := r.needsRemediation(t)
+		if err != nil {
+			log.Error(err, "Reconciling: failed health checking")
+			errList = append(errList, err)
+			continue
+		}
+
+		if needsRemediation {
+			needRemediation = append(needRemediation, t)
+			continue
+		}
+
+		if requeueIn > 0 {
+			log.Info("Reconciling: is likely to go unhealthy", "next check", requeueIn)
+			nodeName := "unknown"
+			if t.Node != nil {
+				nodeName = t.Node.GetName()
+			}
+			r.Recorder.Eventf(t.Node, corev1.EventTypeNormal, utils.EventReasonDetectedUnhealthy,
+				"Machine %v has unhealthy node %v", t.Machine.GetName(), nodeName,
+			)
+			overallRequeueIn = *utils.MinRequeueDuration(&overallRequeueIn, &requeueIn)
+			continue
+		}
+
+		healthy = append(healthy, t)
+	}
+
+	return
+}
+
+func (r *MachineHealthCheckReconciler) needsRemediation(t resources.Target) (bool, time.Duration, error) {
```

Comment:
The meaning of the `time.Duration` return value is unclear unless looking at the code (e.g. line 340) or where the function is used. It might be better to add a short description even if the function is not exported.